open-craft/openedx-completion-aggregator
# [BB-4013] Bump Python to 3.8, upgrade dependencies

**Description:** This upgrades the Python version to 3.8 and bumps constraints to run tests with the dependencies that are used with Koa.
**Note:** The code review should be done commit-by-commit, as they explain why these changes have been made. Also, you don't need to meticulously check `requirements/*.txt` files.

**JIRA:** [BB-4013](https://tasks.opencraft.com/browse/BB-4013)

**Dependencies:** Test instructions require working Koa and Juniper devstacks (any branch).

**Merge deadline:** Before the end of the sprint.

**Installation instructions:** Install this with `pip -e` and restart the container.

**Testing instructions:**
1. See that the CI is passing.
1. Run LMS migrations with `/edx/bin/edxapp-migrate-lms` (this might throw an exception in Juniper - see a note at the bottom).
1. Enable the completion tracking in LMS with `/edx/bin/manage.edxapp lms waffle_switch --create completion.enable_completion_tracking on`.
1. Visit [this page](http://localhost:18000/completion-aggregator/v1/course/course-v1:edX+DemoX+Demo_Course/) and see that it's not throwing any errors.
1. Visit [this subsection](http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/basic_questions/?activate_block_id=block-v1%3AedX%2BDemoX%2BDemo_Course%2Btype%40sequential%2Bblock%40basic_questions) and start completing the blocks. See that the values of the API endpoint (from the previous point) are changing.
1. Complete some blocks from another subsection (e.g. [this one](http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/?activate_block_id=block-v1%3AedX%2BDemoX%2BDemo_Course%2Btype%40sequential%2Bblock%4019a30717eff543078a5d94ae9d6c18a5)) and see that they are included in [this API response](http://localhost:18000/completion-aggregator/v1/course/?requested_fields=sequential).

**Reviewers:**
- [x] @janimo 

**Merge checklist:**

**Post merge:**
      finished.

**Author concerns:** This didn't introduce any breaking changes to Juniper, hence the MINOR bump.

**Juniper dependencies:** While running migrations in Juniper, you might encounter the `django.db.utils.OperationalError: (1060, "Duplicate column name 'banner_image_url'")` exception. If you do, run `/edx/bin/manage.edxapp dbshell` and paste the following code there:
```sql
-- Update the django migration file name
UPDATE django_migrations
    SET name = "0023_courseoverview_banner_image_url" 
    WHERE app = "course_overviews" AND name = "0023_auto_20200922_1245";
 
-- Set default value for courseoverview
ALTER TABLE course_overviews_courseoverview MODIFY banner_image_url longtext DEFAULT '' NOT NULL;
ALTER TABLE course_overviews_historicalcourseoverview MODIFY banner_image_url longtext DEFAULT '' NOT NULL;
```
